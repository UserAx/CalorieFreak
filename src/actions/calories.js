import database from '../firebase/firebase';

//To add new calories record
export const addCalorie = (calorie) => ({
    type: 'ADD_CALORIE',
    calorie
});

export const startAddCalorie = (calorieData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            foodname = '',
            servings = '',
            caloriesAmount = 0,
            createdAt = 0,
            description = ''
        } = calorieData;
        const calorie = {foodname, servings, caloriesAmount, createdAt, description};
        return database.ref(`users/${uid}/calories`).push(calorie).then((ref) => {
            dispatch(addCalorie({
                   id: ref.key,
                    ...calorie
                }));
            });
        }
}

//To remove calories record
export const removeCalorie = ({id} = {}) => ({
    type: 'REMOVE_CALORIE',
    id
});

export const startRemoveCalorie = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/calories/${id}`).remove().then(() => {
            dispatch(removeCalorie({id}));
        });
    };
};

export const startEditCalorie = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/calories/${id}`).update(updates).then(() => {
            dispatch(editCalorie(id, updates));
        });
    };
};

//To edit a calories record
export const editCalorie = (id, updates) => ({
    type: 'EDIT_CALORIE',
    updates,
    id
});

export const startSetCalories = ( ) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/calories`).once('value').then((snapshot) => {
        const calories = [];
        snapshot.forEach((childSnapshot) => {
          calories.push({id: childSnapshot.key,
                        ...childSnapshot.val()});
        });
        dispatch(setCalorie(calories));
      });
    }
}

//To get a calories record from firebase to display on the page
export const setCalorie = (calories) => ({
    type: 'SET_CALORIES',
    calories
});