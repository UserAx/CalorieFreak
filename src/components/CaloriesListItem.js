import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const CaloriesListItem = ({ id, foodname, caloriesAmount, createdAt }) => (
    <Link className = "calorieslistitemlink" to={`edit/${id}`}>
        <div>
            <h2 className = "calorieslistitem__title">{foodname}</h2> 
            <h5 className = "calorieslistitem__date">{moment(createdAt).format('MMMM Do, YYYY')}</h5>
        </div>
        <span className = "calorieslistitem__calories">{caloriesAmount/ 100}</span>
    </Link>
);

export default CaloriesListItem;