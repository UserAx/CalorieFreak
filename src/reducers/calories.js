const calorieDefaultState = [];

export default (state = calorieDefaultState, action) => {
    switch(action.type){
        case 'ADD_CALORIE': 
        return [
            ...state,
            action.calorie
        ];
        case 'SET_CALORIES': 
        return action.calories;
        case 'REMOVE_CALORIE':
        return state.filter(({id}) => id !== action.id);
        case 'EDIT_CALORIE':
        return state.map((calorie) => {
            if(calorie.id === action.id){
                return {
                    ...calorie,
                    ...action.updates
                }
            }else {
                return calorie;
            }
        });
        default:
        return state;
    }
}