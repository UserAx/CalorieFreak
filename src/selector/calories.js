import moment from 'moment';

export default (calories, { text, sortBy, startDate, endDate}) => {
    return calories.filter((calorie) => {
        const createdAtMoment = moment(calorie.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = calorie.foodname.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'calories'){
            return a.caloriesAmount < b.caloriesAmount ? 1 : -1;
        }else if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }
    });
}