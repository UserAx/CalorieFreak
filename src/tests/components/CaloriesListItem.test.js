import React from 'react';
import {shallow} from 'enzyme';
import CaloriesListItem from '../../components/CaloriesListItem';
import calories from '../fixtures/calories';

test('should render calories list item correctly', () => {
    const wrapper = shallow(<CaloriesListItem {...calories[0]} />);
    expect(wrapper).toMatchSnapshot();
});