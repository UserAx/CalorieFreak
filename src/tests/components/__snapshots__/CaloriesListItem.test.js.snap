// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render calories list item correctly 1`] = `
<Link
  to="edit/1"
>
  <div>
    <h2>
      Pasta
    </h2>
    <span>
      150.2425
    </span>
    <h5>
      January 1st, 1970
    </h5>
  </div>
</Link>
`;
